[Woo POS] [Variable Products] Use pagination headers for loading products

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14699
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This pull request introduces a new feature to handle response headers along with response data in network requests. It also includes changes to support paginated responses in the products loading functionality in POS. The most important changes include adding a new method to fetch response data and headers, updating the `ProductsRemote` class to handle paginated responses, and modifying relevant tests to accommodate these changes using a new async version of the pagination tracker.

### Network response handling:

* [`Networking/Networking/Network/AlamofireNetwork.swift`](diffhunk://#diff-e5de4f95ce59918843a27539494c0fe4cb56643666b6ba692a8598eadbb1a35cR87-R102): Added a new method `responseDataAndHeaders` to fetch both data and headers for a network request.
* [`Networking/Networking/Network/MockNetwork.swift`](diffhunk://#diff-b3daa535a62ca6de09b8f4077bc87980d04ee9e9e3dce0bda93e43fccd5b5d1bR30-R32): Added `responseDataAndHeaders` method and `responseHeaders` variable to mock network responses with headers. [[1]](diffhunk://#diff-b3daa535a62ca6de09b8f4077bc87980d04ee9e9e3dce0bda93e43fccd5b5d1bR30-R32) [[2]](diffhunk://#diff-b3daa535a62ca6de09b8f4077bc87980d04ee9e9e3dce0bda93e43fccd5b5d1bR84-R97)
* [`Networking/Networking/Network/Network.swift`](diffhunk://#diff-1241b41f6051bf98c4b482192989aee19ca092ba431b5447ea26da1bcf814f8eR21): Introduced `ResponseHeaders` typealias and added `responseDataAndHeaders` method to the `Network` protocol. [[1]](diffhunk://#diff-1241b41f6051bf98c4b482192989aee19ca092ba431b5447ea26da1bcf814f8eR21) [[2]](diffhunk://#diff-1241b41f6051bf98c4b482192989aee19ca092ba431b5447ea26da1bcf814f8eR43-R44)
* [`Networking/Networking/Network/NullNetwork.swift`](diffhunk://#diff-7e68cebf68a33ece6b704356f4e51455a456a905b72212595978af4534556982R22-R25): Implemented `responseDataAndHeaders` method to throw `NetworkError.notFound` as a placeholder implementation.
* [`Networking/Networking/Network/WordPressOrgNetwork.swift`](diffhunk://#diff-826ad9712487a498a6ea4a20fc0a240bfa7401526cceaa4410c1679f3526ff23R107-R122): Added `responseDataAndHeaders` method to handle response data and headers in unit tests.

### Paginated responses:

* [`Networking/Networking/Remote/ProductsRemote.swift`](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6L207-R209): Updated `loadProductsForPointOfSale` method to return `PagedItems` and handle pagination using response headers. [[1]](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6L207-R209) [[2]](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6L225-R236)
* [`Networking/Networking/Remote/Remote.swift`](diffhunk://#diff-867c8f6d7c2e8dd17faadda0483244aaf07ae31c1afecc9af3878a093d84d33bR261-R287): Added `enqueueWithResponseHeaders` method to handle responses with headers and introduced `PagedItems` struct to represent paginated results. [[1]](diffhunk://#diff-867c8f6d7c2e8dd17faadda0483244aaf07ae31c1afecc9af3878a093d84d33bR261-R287) [[2]](diffhunk://#diff-867c8f6d7c2e8dd17faadda0483244aaf07ae31c1afecc9af3878a093d84d33bR369-R379) [[3]](diffhunk://#diff-867c8f6d7c2e8dd17faadda0483244aaf07ae31c1afecc9af3878a093d84d33bR388-R391)

### Pagination Improvements:
* [`WooCommerce/Classes/Tools/InfiniteScroll/AsyncPaginationTracker.swift`](diffhunk://#diff-fedbcd04a4fa1813ce8096c04245c0cfd6ebdb553891e153ce286062ab9a99fdR1-R131): Introduced the `AsyncPaginationTracker` class to manage pagination state and support infinite scroll and pull-to-refresh. Most of the implementation is based on the closure-based `PaginationTracker`.
* [`WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift`](diffhunk://#diff-b34ef3931ae78bf4b51520edbe9c11abd6bf36d31dfce4882f3b621c5a7e2af9L18-R88): Refactored to use `AsyncPaginationTracker` for managing pagination state.

### Tests:

* [`Networking/NetworkingTests/Remote/ProductsRemoteTests.swift`](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL943-R946): Updated tests to handle `PagedItems` and verify pagination logic. [[1]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL943-R946) [[2]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL973-R977) [[3]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL991-R1032)

### Additional changes:

* [`WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift`](diffhunk://#diff-b9d72a7102bb1a5778e338bc3b2046cd3490aa8dc893fc3a05f18cb655ce698cR64): Added a task to manage item reloads. [[1]](diffhunk://#diff-b9d72a7102bb1a5778e338bc3b2046cd3490aa8dc893fc3a05f18cb655ce698cR64) [[2]](diffhunk://#diff-b9d72a7102bb1a5778e338bc3b2046cd3490aa8dc893fc3a05f18cb655ce698cR104-R109)
* [`WooCommerce/Classes/POS/Utils/PreviewHelpers.swift`](diffhunk://#diff-97e03cb12a8d584183bc71ac17d2034fc198adb981132042599469d40e9c8b3dR11): Updated mock service to return `PagedItems`. [[1]](diffhunk://#diff-97e03cb12a8d584183bc71ac17d2034fc198adb981132042599469d40e9c8b3dR11) [[2]](diffhunk://#diff-97e03cb12a8d584183bc71ac17d2034fc198adb981132042599469d40e9c8b3dL40-R42)

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

- Switch to a store eligible for POS
- Go to Menu > Point of Sale mode --> the products should be loaded after the initial loading state
- Scroll toward the bottom --> if the store has more than 100 simple/variable published products, the next page of products should be loaded with a placeholder row at the bottom

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

- [x] @jaclync tests on stores with more than 1 page of products, and just 1 page of products.
- [ ] @jaclync tests on stores with site credentials login

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/eff17078-10fc-4932-9e90-0a15d28fb14e



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

jaclync committed:
WIP: first working version of returning response data and headers from remote, get the total pages count from header if available, then determine if it has next page. Replace 2 internal pagination states with `PaginationTracker`.

jaclync committed:
Update network implementation to use Alamofire's sessionRequest.serializingData async function.

jaclync committed:
Fix initialLoading state is gone while items are being loaded.

jaclync committed:
Fix `ProductsRemoteTests` existing test failures.

jaclync committed:
Add a test case on page count value from case-insensitive header.

jaclync committed:
Create `AsyncPaginationTracker` based on `PaginationTracker` to enable async/await interface in POS. Some PointOfSaleItemsControllerTests cases are still failing.

jaclync committed:
Update test cases in `PointOfSaleItemsControllerTests` based on the expected behavior and the move of pagination tracking to PaginationTracker with fixes for issues found in the tests.

jaclync committed:
Fix `PointOfSaleProductService` test case for special case when empty data are returned for page number > 1.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jaclync committed:
Fix SwiftLint error.

jaclync commented on Yosemite/Yosemite/PointOfSale/PointOfSaleProductService.swift:
```diff
@@ -41,13 +40,13 @@ public final class PointOfSaleProductService: PointOfSaleItemServiceProtocol {
 
         if pageNumber != 1 && products.count == 0 {
-            throw PointOfSaleProductServiceError.pageOutOfRange
+            return ([], false)
```
🗒️ for reviewer: this case should not be reached if the pagination headers are available. When the pagination headers are missing, this "page out of range" case indicates that there there are no more data to load and thus return empty items and `hasNextPage: false`.

joshheald replied:
Thanks. If we want to keep this approach (rather than throwing when the header isn't available/an int) then let's add this comment in the code :)

I think it would also be reasonable to make the headers required though. All tunnelled requests will have them, and I think it's a really old header. We could check whether it's possible to have a site-credential connection without getting these headers back – first step would be to check WC3.5 as I think that's the minimum.

jaclync commented on WooCommerce/WooCommerceTests/POS/Controllers/PointOfSaleItemsControllerTests.swift:
```diff
@@ -125,6 +125,7 @@ final class PointOfSaleItemsControllerTests {
         let initialItems = MockPointOfSaleItemService.makeInitialItems()
         itemProvider.items = initialItems
         itemProvider.shouldSimulateTwoPages = true
+        await sut.loadInitialItems()
```
🗒️ for reviewer: `sut.loadInitialItems()` was added in a few test cases before `sut.loadNextItems()`, because the latest page number tracking has been moved from the items controller to the pagination tracker `AsyncPaginationTracker`. Before this move, the page number is already 1 by default in the items controller without actually having fetched the first page. When calling `loadNextItems`, it increments the page number to 2 and the mock item provider appends the page 2 items to the page 1 items and returns all these items for the page 2 request:

https://github.com/woocommerce/woocommerce-ios/blob/931208508fe47dfffccd537a36698abd8d07c50f/WooCommerce/WooCommerceTests/POS/Mocks/MockPOSItemProvider.swift#L34-L36

This behavior doesn't seem expected, as the item provider just returns the items for a given page number without any state from previous pages. This PR removes this behavior, but please let me know if I misunderstood the expected behavior of item provider.

joshheald replied:
Yeah that seems better

jaclync commented on WooCommerce/WooCommerceTests/POS/Controllers/PointOfSaleItemsControllerTests.swift:
```diff
@@ -229,33 +242,26 @@ final class PointOfSaleItemsControllerTests {
         #expect(itemProvider.spyLastRequestedPageNumber == 1)
     }
 
-    @Test func loadNextItems_when_next_page_is_out_of_range_then_receives_error() async throws {
+    @Test func loadNextItems_when_next_page_is_empty_then_state_is_loaded() async throws {
```
🗒️ for reviewer: similar to another comment https://github.com/woocommerce/woocommerce-ios/pull/14728/files#r1891180720, the "page out of range" error was removed because receiving empty data in > 1 page number is how we tell we've reached the last page when pagination headers are unavailable. In this scenario, I think it makes sense to transition the state to `loaded`.

joshheald replied:
Yep, agreed. The "error" was a hack so that we could figure out when we'd gone too far and should stop requesting extra pages – it shouldn't be needed now that we have the total pages header.

jaclync committed:
Update comments in `AsyncPaginationTracker`.

jaclync committed:
Update `AsyncPaginationTracker` to mention the renaming.

jaclync committed:
Fix reload task being canceled when releasing the refresh control, because `refreshable` action gets canceled on release.

joshheald commented on Networking/Networking/Network/Network.swift:
```diff
@@ -18,6 +18,7 @@ public protocol MultipartFormData {
 /// Unit Testing target, and inject mocked up responses.
 ///
 public protocol Network {
+    typealias ResponseHeaders = [String: String]
```
Is it worth adding a `ResponseHeader` struct and keeping them in an array?

jaclync replied:
Right now it's a dictionary for O(1) lookup, did you mean to convert it to convert the key-value pairs to an array? Or having a struct with a dictionary field?

joshheald replied:
I meant something like
```
struct ResponseHeader {
    let name: String
    let value: String
}

public protocol Network {
    func responseDataAndHeaders(for request: URLRequestConvertible) async throws -> (Data, [ResponseHeader])
}
```

Not a strong requirement, just wondered whether you thought it would make the code clearer. I didn't really consider performance, but the array should be small.

jaclync replied:
Personally, I feel like `[String: String]` is pretty similar to the key-value pairs array both with the need to look up and optional return value, and with guaranteed performance. Maybe we can keep it simple with a dictionary for now, and update to a struct if we have more use cases of the headers.

joshheald commented on Networking/Networking/Remote/ProductsRemote.swift:
```diff
@@ -222,7 +224,15 @@ public final class ProductsRemote: Remote, ProductsRemoteProtocol {
+        // Extracts the total number of pages from the response headers.
+        // Response header names are case insensitive.
+        let totalPages = responseHeaders?.first(where: { $0.key.lowercased() == Remote.PaginationHeaderKey.totalPagesCount.lowercased() })
+            .flatMap { Int($0.value) }
```
Should this use `compactMap`?

Should we throw if it's not there, rather than make it optional? I don't know how well we'll be able to handle the list if we don't have the page count...

jaclync replied:
Hm good question, I was mostly afraid that some plugins or hosts (for site credentials login with direct site requests) might alter the headers and remove the pagination headers. In this case where the pagination headers are missing, I'm not sure if it's worth blocking the merchants from using POS because of the missing headers. WDYT?

joshheald replied:
I think it depends... 
1. What complexity does any workaround add to the code? (If we have to keep the previous workaround in, that's a strong sign we shouldn't do it.)
2. What kind of experience will we give them if the headers aren't present? (If it's just going to keep trying to load the next page, that's not great either, especially if those calls are rapidly repeated which is a risk here.)
3. How likely is it that the headers aren't present?

[The docs for v3 of the API say:](https://woocommerce.github.io/woocommerce-rest-api-docs/#pagination)
> The total number of resources and pages are always included in the X-WP-Total and X-WP-TotalPages HTTP headers.
[Which is repeated back to v1 of the API](https://woocommerce.github.io/woocommerce-rest-api-docs/wp-api-v1.html#pagination) – WC 2.6 got that, and we don't support that version. 

I _think_ we can rely on them being there, but perhaps we could add some tracking now for any request that comes back without the headers, if you have concerns?

jaclync replied:
> What complexity does any workaround add to the code? (If we have to keep the previous workaround in, that's a strong sign we shouldn't do it.)

Great points on the various considerations and tradeoffs. I think the current workaround is not very complicated even though not ideal - it's mostly in the [remote `loadProductsForPointOfSale` returning true](https://github.com/woocommerce/woocommerce-ios/blob/cc139f83809727c06f69264306f252cd1957b1d9/Networking/Networking/Remote/ProductsRemote.swift#L234) for `hasMorePages` by default, and [`PointOfSaleProductService` returning `false` when the page items are empty](https://github.com/woocommerce/woocommerce-ios/blob/cc139f83809727c06f69264306f252cd1957b1d9/Yosemite/Yosemite/PointOfSale/PointOfSaleProductService.swift#L49-L51) (there's room for optimization there to return false for items count < page size). Having dealt with all kinds of alterations in the API response by third-party plugins, I don't know about the likelihood but there's a chance some plugin or hosting provider strips some response headers. Of course there could be none, we can add tracking to the response headers missing case and observe. If it's almost zero, let's move to throw an error when response headers are missing to simplify the code. I created an issue to track this https://github.com/woocommerce/woocommerce-ios/issues/14737.

> What kind of experience will we give them if the headers aren't present? (If it's just going to keep trying to load the next page, that's not great either, especially if those calls are rapidly repeated which is a risk here.)

There should be up to 1 extra API request per load session with the current fallback implementation (before the next reload), when we fix the infinite scroll issue. If we implement the optimization to return false for response items count < page size, then the probability of extra API request is lower (only if the last page has items size = page size).

joshheald commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -15,70 +15,76 @@ class PointOfSaleItemsController: PointOfSaleItemsControllerProtocol {
 
     init(itemProvider: PointOfSaleItemServiceProtocol) {
         self.itemProvider = itemProvider
+        self.paginationTracker = .init(pageFirstIndex: Constants.initialPage)
```
I think you can remove the constant here and just use the default initialiser

jaclync replied:
Makes sense, updated in https://github.com/woocommerce/woocommerce-ios/pull/14728/commits/27e2f91706060133d98f77350c0eedba52aab2be

joshheald commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -15,70 +15,76 @@ class PointOfSaleItemsController: PointOfSaleItemsControllerProtocol {
+            switch nextPageState {
+                case .noNextPage, .synced:
+                    updateItemListStateAfterLoadAttempt()
+                case .syncing:
+                    break
```
How would we get here while it was syncing? Is it impossible?

jaclync replied:
If `loadNextItems` is called again while the previous call is still loading items remotely, it reaches this state.

joshheald replied:
With the suggestion to gate calls to `loadNextItems`, that _shouldn't_ happen, but it's good to handle it in case the caller isn't doing that.

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -100,7 +101,10 @@ extension PointOfSaleAggregateModel {
+        reloadTask?.cancel()
+        reloadTask = Task { @MainActor in
+            await itemsController.reload()
+        }
```
Good catch... though could/should this be inside the itemsController?

jaclync replied:
Yup better not clutter the aggregate model, moved in https://github.com/woocommerce/woocommerce-ios/pull/14728/commits/2d3480697db6157521f0fadb61664da5ebac6e10

joshheald replied:
I'm looking a bit closer at this now that it's breaking our unit tests... could you explain again what the problem was here? Was the action cancelled when you let go of the control?

That shouldn't really happen – the `refreshable` modifier accepts an async function and should work with it... I'd expect it to cancel the task if the view goes off screen. Breaking the async chain by dispatching to a task seems like it causes issues – when you release the list the spinner disappears immediately becaus the `reload` function returns before the network response comes back

joshheald replied:
Hmmm I think I can see what's going on now. The error we're getting is `explicitlyCancelled` – it looks like we're doing that by causing the view to redraw within the reload function. We call `itemListStateSubject.send(.loading(allItems))` when it's called, which redraws the view to show the loading state at the top... but doing that redraw means that SwiftUI cancels the refresh control.

If we remove that line, the refresh works as you'd expect, but without the loading cell at the top. To be honest, I think that's a better trade off for the time being.

joshheald replied:
The problem is that we don't have fine-grained state updates yet – every time we update the POSModel we redraw everything. 

I think we would be permitted to change the list _content_ without the refresh action getting cancelled, but because we don't have `Observation` yet our changes result in the list itself being redrawn, including the refresh control.

jaclync replied:
Yes the `explicitlyCancelled` cancelation that results in error state is the issue. Thanks for the better solution by not triggering redraw in https://github.com/woocommerce/woocommerce-ios/pull/14728/commits/1fb8df061cb1965ef071aa008aa63c49a17c26c8! From my testing with stores with 2 POS products in your PR https://github.com/woocommerce/woocommerce-ios/pull/14727, I occasionally still saw this error but only briefly because the pre-existing issue https://github.com/woocommerce/woocommerce-ios/issues/14735 where `loadNextItems` is also triggered when pulling to refresh a short list (this request isn't canceled and the items from this request update the items view state). Would be great to use `Observation` to fix this issue properly.

jaclync committed:
Update `ProductsRemote.loadProductsForPointOfSale` to return `PagedItems` that contains both the items and whether there are more pages.

joshheald commented on WooCommerce/Classes/Tools/InfiniteScroll/AsyncPaginationTracker.swift:
```diff
@@ -0,0 +1,132 @@
+    ///     1.  Proceed only if there is next page to sync.
+    ///     2.  Verify if the next page isn't currently being synced.
+    ///     3.  Proceed syncing the next page.
+    func ensureNextPageIsSynced(syncFunction: @escaping SyncFunction) async throws -> NextPageSyncState {
```
Did you consider/try passing the `syncFunction` in through `init`? 

At the call site, the repeated trailing closures feel repetitive, at least for the current example. As long as the sync functions we call are `async throws`, we can set our state accordingly before and after the calls from ItemListController.

jaclync replied:
I considered passing the sync function through init, but the sync involves states in self (`allItems` at least) so it cannot be passed in the initializer. I agree it feels repetitive passing for each method. We could consider updating this function to be static functional without any states, but that requires some refactoring and I'd prefer leaving it for later.

joshheald replied:
Ah yeah, of course. OK, let's keep it like this for now, I don't think it'll be realistic to make it static, because then the consumer will be really limited about what it can do with the state it stores, and testing will be a pain.

joshheald commented on WooCommerce/Classes/Tools/InfiniteScroll/AsyncPaginationTracker.swift:
```diff
@@ -0,0 +1,132 @@
+        markAsBeingSynced(pageNumber: pageNumber)
+
+        do {
+            defer {
+                unmarkAsBeingSynced(pageNumber: pageNumber)
+            }
```
I had trouble with defer working as I expected when I used it like this in the ItemsListController – are you sure it gets called if we throw an error? Should it be?

jaclync replied:
Yes, if this defer statement is commented out, `loadNextItems_after_itemProvider_throws_error_then_the_same_page_is_requested_next` fails from the page not being marked as synced. `unmarkAsBeingSynced` was originally called after the do/catch, but when an error is thrown it's never reached. Moving it to `defer` gets called in both cases.

From the [Apple doc](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements/#Defer-Statement):

> The statements within the defer statement are executed no matter how program control is transferred. This means that a defer statement can be used, for example, to perform manual resource management such as closing file descriptors, and to perform actions that need to happen even if an error is thrown.

joshheald replied:
Great, thanks for checking. Goodness knows what I was doing wrong before then... perhaps I accidentally had it in the wrong scope or something.

jaclync committed:
Update `PointOfSaleItemServiceProtocol.providePointOfSaleItems` to also return `PagedItems`.

joshheald reviewed (commented):
Thanks for the async handler!

I haven't tested any of this yet, just some thoughts so far.

joshheald commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -79,12 +79,13 @@ class PointOfSaleItemsController: PointOfSaleItemsControllerProtocol {
         }
         allItems.append(contentsOf: uniqueNewItems)
-        return hasNextPage
+        return pagedItems.hasMorePages
```
It's been there a while... but this function is feeling a bit strange now.

It's called `fetchItems`, which it does, but it doesn't return them, it returns a Bool.

If I didn't have the doc comment, I'd expect that to indicate success/failure of the fetch, not whether there are more pages. It's not a huge deal on this private function but it's niggling at me a bit.

We can leave it for now, this one will change more when the state model changes, but it'd be good to know your thoughts. I guess just returning the whole `PagedItems` actually would make more sense...

jaclync replied:
This is the function called from the pagination tracker to sync items given a page number, which just requires a boolean return value that indicates whether there are more pages. Its usage/meaning does feel unintuitive without the doc comment.

I think it looks awkward because the items and pagination state are stored separately, but both came from one network request. I tried making the sync function stateless as in https://github.com/woocommerce/woocommerce-ios/pull/14728#discussion_r1891526797 by passing the sync function `try await itemProvider.providePointOfSaleItems(pageNumber: $0).hasMorePages` in init, but then the controller cannot get the items easily. I'll think more about how the items and pagination states can be managed more cleanly, feel free to share any ideas as well.

joshheald replied:
Other than changing the function to return `PagedItems<POSProduct>`, which includes both the items and whether there are pages, I don't have any ideas.

Let's keep it this way; a better one might become obvious after we use it for variations.

joshheald commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -15,70 +15,77 @@ class PointOfSaleItemsController: PointOfSaleItemsControllerProtocol {
 
     @MainActor
     func loadNextItems() async {
+        itemListStateSubject.send(.loading(allItems))
```
```suggestion
        guard paginationTracker.hasNextPage else {
            return
        }
        itemListStateSubject.send(.loading(allItems))
```

You'll also need to make `hasNextPage` a `private(set)` var as well.

jaclync replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/14728/commits/a2e792aff49ce36f2ea69078bd6839dfe33666ad, this should fix the flickering issue when loading the next page from the infinite scroll.

joshheald reviewed (approved):
With testing, it works quite well. There are some issues but they're sort of preexisting...

1. There's a flickering of the loading next page state when we get to the bottom of the last page. This previously didn't happen, because there was a check in ItemListController before sending the `loading` state. It's sort of pre-existing though because it's caused by an issue with the load next items gesture, which fires ~50 times when you scroll at the bottom of the page. I've suggested a change which hides it again.
2. Uniquing the items doesn't work at the moment – this is because of the UUIDs, and nothing to do with this PR.

Taken with any changes from my previous review, I think this is fine to merge,

jaclync committed:
Use default first page index.

jaclync committed:
Fix crashed unit tests from `MockPointOfSaleItemService` not having the correct return type.

jaclync committed:
Move separate reload task from aggregate model to items controller.

jaclync committed:
When loading next items, only show the loading state and continue with loading the next page if there are more pages from the pagination tracker.

jaclync commented:
@joshheald Thanks for the feedback! I believe I've responded to all of your comments, and the flickering issue is hopefully fixed in https://github.com/woocommerce/woocommerce-ios/pull/14728#discussion_r1892057124 if you'd like to test it again.

> Uniquing the items doesn't work at the moment – this is because of the UUIDs, and nothing to do with this PR.

Did this cause any visual issues in your testing?

joshheald commented:
> > Uniquing the items doesn't work at the moment – this is because of the UUIDs, and nothing to do with this PR.
> 
> Did this cause any visual issues in your testing?

Yes – after errors on fetching the next page, I think I was able to get duplicate items in the list. It's nothing to do with this PR though, we should fix it separately.

I'll check it again now

jaclync commented:
The move of the reload task in https://github.com/woocommerce/woocommerce-ios/pull/14728#discussion_r1892062784 broke the reload test cases 😬 Looking into them now.

joshheald commented:
@jaclync some of the unit test failures suggest that the async code is returning before it's finished its work somewhere... there are a few failing because they're expecting a loaded state but getting a previous loading one....

jaclync commented:
> some of the unit test failures suggest that the async code is returning before it's finished its work somewhere... there are a few failing because they're expecting a loaded state but getting a previous loading one....

Yea, because I moved the separate reload task for the refresh control fix from the aggregate model to the items controller. I will make a separate function to use the separate reload Task, and have unit tests use the core reload function.

jaclync committed:
Move reload core function to a separate function for testing - not ideal, but other places to separate the reload task are the SwiftUI view (which cannot mutate self in refreshable) or the aggregate model have their downsides as well. Having to wait for the conditions in all 5 test cases makes it look hacky.

jaclync commented:
The commit https://github.com/woocommerce/woocommerce-ios/pull/14728/commits/4db636839f61994aa3e8dc22cc2ae819ad561855 fixes the test failures by moving the core reload function to a separate function, and have unit tests use the other reload function. This isn't ideal because this function doesn't have to be public. When I have some time this/next week, I'll look into how to wait for conditions better like the XCTest `waitUntil` helper. I just didn't want to add the verbose wait workaround for 5 failing test cases in a haste.

joshheald committed:
Revert "Move reload core function to a separate function for testing - not ideal, but other places to separate the reload task are the SwiftUI view (which cannot mutate self in refreshable) or the aggregate model have their downsides as well. Having to wait for the conditions in all 5 test cases makes it look hacky."

This reverts commit 4db636839f61994aa3e8dc22cc2ae819ad561855.

joshheald committed:
14699 Avoid changing the view state during refresh

Previously we sent a `.loading` state in the pull to refresh handler.

That changes the POSModel observable object, which results in the whole POS being redrawn.

Redrawing the refresh controller cancels the ongoing async task so it always shows an error.

With this change, we lose the skeleton loading cell when pulling to refresh, but the function works and the default control still shows.

With more granular state observation, we could update the list content without redrawing the list component itself, avoiding this issue.
